contributing/topics/guide-new-resource.md

```diff
@@ -637,6 +653,128 @@ func (ResourceGroupExampleResource) expandComplexResource(input []ComplexResourc
+
+For example, in this case:
+
+:white_check_mark: **DO**
+
+```
+"name": {
+	Type:     pluginsdk.TypeString,
+	Required: true,
+	ForceNew: true,
+	ValidateFunc: validation.StringMatch(
+		regexp.MustCompile("^[a-zA-Z0-9]([a-zA-Z0-9-_]{0,78}[a-zA-Z0-9])?$"),
+		"The `name` can only contain alphanumeric characters, underscores and dashes up to 80 characters in length.",
+	),
+},
+
+return fmt.Errorf("retrieving %s: `model` was nil", id)
+return fmt.Errorf("retrieving %s: `properties` was nil", id)
+```
+
+:no_entry: **DO NOT**
+
+```
+"name": {
+	Type:     pluginsdk.TypeString,
+	Required: true,
+	ForceNew: true,
+	ValidateFunc: validation.StringMatch(
+		regexp.MustCompile("^[a-zA-Z0-9]([a-zA-Z0-9-_]{0,78}[a-zA-Z0-9])?$"),
+		"The name can only contain alphanumeric characters, underscores and dashes up to 80 characters in length.",
+	),
+},
+
+return fmt.Errorf("retrieving %s: model was nil", id)
+return fmt.Errorf("retrieving %s: properties was nil", id)
+```
+
+- Do not return an error in `Update` or `CustomizeDiff` function if the error is by design and cannot be resolved by changing the configuration, use `ForceNew` in `CustomizeDiff` instead so that Terraform will plan a resource recreation.
+
+:white_check_mark: **DO**
+
+```go
+func (r ExampleResource) CustomizeDiff() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			...
+			o, n := metadata.ResourceDiff.GetChange("zone_balancing_enabled")
+			if o.(bool) != n.(bool) {
+				// Changing `zone_balancing_enabled` from `false` to `true` requires the capacity of the sku to be greater than `1`.
+				if !o.(bool) && n.(bool) && rd.Get("worker_count").(int) < 2 {
+					if err := metadata.ResourceDiff.ForceNew("zone_balancing_enabled"); err != nil {
+						return err
+					}
+				}
+			}
+
+			...
+
+			return nil
+		},
+	}
+}
+```
+
+:no_entry: **DO NOT**
+
+```go
+func (r ExampleResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 60 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			...
+            o, n := metadata.ResourceData.GetChange("zone_balancing_enabled")
+			if o.(bool) != n.(bool) {
+				if !o.(bool) && n.(bool) && rd.Get("worker_count").(int) < 2 {
+					return errors.New("Changing `zone_balancing_enabled` from `false` to `true` requires the capacity of the sku to be greater than `1`.")
+				}
+			}
+
+			...
+			return nil
+		},
+	}
+}
+```
+
+- `pointer.From` returns the dereferenced value or zero if the pointer is `nil`. Use `pointer.From` instead of manual `nil` checks.
+
+:white_check_mark: **DO**
+
+```go
+output.Name = pointer.From(input.Name)
+```
+
+:no_entry: **DO NOT**
+
+```go
+if input.Name != nil {
+    output.Name = *input.Name
+}
+```
+
+- Use `pointer.ToEnum` to convert Enum type instead of explicitly type conversion.
```

Comment:
Can we also mention when to use `pointer.To`, to make the document comprehensive?

Reply:
added above, thanks.